### PR TITLE
[SEMI-MODULAR] Gunpointing! Similar to bay-aiming

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request
This is pretty much the gunpointing we've had before. Had to add an extra trait and signals and stuff, also the "direct" variable for radios.

## Changelog
:cl:
/:cl:

### CODE REVIEW

File: code/_onclick/hud/radial.dm [hunk 1]
```diff
@@ -122,6 +123,7 @@ GLOBAL_LIST_EMPTY(radial_menus)
 		var/elements_to_add = max_elements - elements.len
 		for(var/i in 1 to elements_to_add) //Create all elements
 			var/obj/screen/radial/slice/new_element = new /obj/screen/radial/slice
+			new_element.icon = icon_path //SKYRAT CHANGE - module GUNPOINT
```
Comment (Gandalf2k15) [inline code/_onclick/hud/radial.dm]: //SKYRAT EDIT ADDITION - GUNPOINT

File: code/_onclick/hud/radial.dm [hunk 2]
```diff
@@ -80,6 +80,7 @@ GLOBAL_LIST_EMPTY(radial_menus)
 	var/hudfix_method = TRUE //TRUE to change anchor to user, FALSE to shift by py_shift
 	var/py_shift = 0
 	var/entry_animation = TRUE
+	var/icon_path = 'icons/mob/radial.dmi' //SKYRAT CHANGE - module GUNPOINT
```
Comment (Gandalf2k15) [inline code/_onclick/hud/radial.dm]: //SKYRAT EDIT ADDITION - GUNPOINT

File: code/_onclick/item_attack.dm [hunk 3]
```diff
@@ -23,6 +23,7 @@
 /obj/item/proc/attack_self(mob/user)
 	if(SEND_SIGNAL(src, COMSIG_ITEM_ATTACK_SELF, user) & COMPONENT_NO_INTERACT)
 		return
+	SEND_SIGNAL(user, COMSIG_MOB_ITEM_ATTACK_SELF, src) //SKYRAT EDIT - module GUNPOINT - extra signal
```
Comment (Gandalf2k15) [inline code/_onclick/item_attack.dm]: //SKYRAT EDIT ADDITION - GUNPOINT

File: code/game/objects/items/devices/radio/headset.dm [hunk 4]
```diff
@@ -61,7 +61,7 @@ GLOBAL_LIST_INIT(channel_tokens, list(
 	QDEL_NULL(keyslot2)
 	return ..()
 
-/obj/item/radio/headset/talk_into(mob/living/M, message, channel, list/spans, datum/language/language, list/message_mods)
+/obj/item/radio/headset/talk_into(mob/living/M, message, channel, list/spans, datum/language/language, list/message_mods, direct = TRUE) //SKYRAT EDIT - module GUNPOINT - adds direct variable
```
Comment (Gandalf2k15) [inline code/game/objects/items/devices/radio/headset.dm]: ///obj/item/radio/headset/talk_into(mob/living/M, message, channel, list/spans, datum/language/language, list/message_mods) - ORIGINAL
/obj/item/radio/headset/talk_into(mob/living/M, message, channel, list/spans, datum/language/language, list/message_mods, direct = TRUE) //SKYRAT EDIT CHANGE - GUNPOINT

File: code/game/objects/items/devices/radio/radio.dm [hunk 5]
```diff
@@ -185,7 +185,7 @@
 					recalculateChannels()
 				. = TRUE
 
-/obj/item/radio/talk_into(atom/movable/M, message, channel, list/spans, datum/language/language, list/message_mods)
+/obj/item/radio/talk_into(atom/movable/M, message, channel, list/spans, datum/language/language, list/message_mods, direct = TRUE) //SKYRAT EDIT - module GUNPOINT - adds direct variable
```
Comment (Gandalf2k15) [inline code/game/objects/items/devices/radio/radio.dm]: ///obj/item/radio/talk_into(atom/movable/M, message, channel, list/spans, datum/language/language, list/message_mods) - ORIGINAL
/obj/item/radio/talk_into(atom/movable/M, message, channel, list/spans, datum/language/language, list/message_mods, direct = TRUE) //SKYRAT EDIT CHANGE - GUNPOINT

File: code/game/objects/items/devices/radio/radio.dm [hunk 6]
```diff
@@ -203,6 +203,7 @@
 		spans = list(M.speech_span)
 	if(!language)
 		language = M.get_selected_language()
+	SEND_SIGNAL(M, COMSIG_MOVABLE_RADIO_TALK_INTO, src, message, channel, spans, language, direct) //SKYRAT EDIT - module GUNPOINT - extra signal
```
Comment (Gandalf2k15) [inline code/game/objects/items/devices/radio/radio.dm]: //SKYRAT EDIT ADDITION - GUNPOINT

File: code/game/objects/items/devices/radio/radio.dm [hunk 7]
```diff
@@ -298,7 +299,7 @@
 			if (idx && (idx % 2) == (message_mods[RADIO_EXTENSION] == MODE_L_HAND))
 				return
 
-	talk_into(speaker, raw_message, , spans, language=message_language)
+	talk_into(speaker, raw_message, , spans, language=message_language, direct=FALSE) //SKYRAT EDIT - module GUNPOINT - differentiate between things spoken into radios directly vs overheard by an intercom/station-bounced
```
Comment (Gandalf2k15) [inline code/game/objects/items/devices/radio/radio.dm]: //	talk_into(speaker, raw_message, , spans, language=message_language)
	talk_into(speaker, raw_message, , spans, language=message_language, direct=FALSE) //SKYRAT EDIT CHANGE - GUNPOINT - differentiate between things spoken into radios directly vs overheard by an intercom/station-bounced

File: code/modules/mob/living/carbon/human/examine.dm [hunk 8]
```diff
@@ -428,6 +428,13 @@
 					"<a href='?src=[REF(src)];hud=s;add_comment=1'>\[Add comment\]</a>"), "")
 	else if(isobserver(user))
 		. += "<span class='info'><b>Traits:</b> [get_quirk_string(FALSE, CAT_QUIRK_ALL)]</span>"
+	//SKYRAT CHANGES BEGIN - module GUNPOINT
```
Comment (Gandalf2k15) [inline code/modules/mob/living/carbon/human/examine.dm]: //SKYRAT EDIT ADDITION BEGIN - GUNPOINT

File: code/modules/mob/living/carbon/human/examine.dm [hunk 9]
```diff
@@ -428,6 +428,13 @@
...
+	if(gunpointing)
+		. += "<span class='warning'><b>[t_He] [t_is] holding [gunpointing.target.name] at gunpoint with [gunpointing.aimed_gun.name]!</b></span>\n"
+	if(length(gunpointed))
+		for(var/datum/gunpoint/GP in gunpointed)
+			. += "<span class='warning'><b>[GP.source.name] [GP.source.p_are()] holding [t_him] at gunpoint with [GP.aimed_gun.name]!</b></span>\n"
+	//SKYRAT CHANGES END
```
Comment (Gandalf2k15) [inline code/modules/mob/living/carbon/human/examine.dm]: //SKYRAT EDIT ADDITION END

File: code/modules/mob/living/living.dm [hunk 10]
```diff
@@ -116,6 +116,22 @@
 					if(!(world.time % 5))
 						to_chat(src, "<span class='warning'>[L] is restraining [P], you cannot push past.</span>")
 					return TRUE
+		//SKYRAT CHANGES START - module GUNPOINT
```
Comment (Gandalf2k15) [inline code/modules/mob/living/living.dm]: //SKYRAT EDIT ADDITION BEGIN - GUNPOINT

File: code/modules/mob/living/living.dm [hunk 11]
```diff
@@ -116,6 +116,22 @@
...
+					break
+			if(!is_pointing)
+				if(!(world.time % 5))
+					to_chat(src, "<span class='warning'>[L] is being held at gunpoint, it's not wise to push him.</span>")
+				return TRUE
+		if(L.gunpointing)
+			if(!(world.time % 5))
+				to_chat(src, "<span class='warning'>[L] is holding someone at gunpoint, you cannot push past.</span>")
+			return TRUE
+		//END OF SKYRAT CHANGES
```
Comment (Gandalf2k15) [inline code/modules/mob/living/living.dm]: //SKYRAT EDIT ADDITION END

File: code/modules/mob/living/living.dm [hunk 12]
```diff
@@ -1282,6 +1298,7 @@
 			add_or_update_variable_movespeed_modifier(/datum/movespeed_modifier/limbless, multiplicative_slowdown = limbless_slowdown)
 		else
 			remove_movespeed_modifier(/datum/movespeed_modifier/limbless)
+	SEND_SIGNAL(src, COMSIG_LIVING_UPDATED_MOBILITY, mobility_flags) //SKYRAT CHANGE - module GUNPOINT - extra signal
```
Comment (Gandalf2k15) [inline code/modules/mob/living/living.dm]: //SKYRAT EDIT ADDITION - GUNPOINT

File: code/modules/mob/mob_movement.dm [hunk 13]
```diff
@@ -449,6 +449,11 @@
 	if(m_intent == MOVE_INTENT_RUN)
 		m_intent = MOVE_INTENT_WALK
 	else
+		//SKYRAT EDIT BEGIN - module GUNPOINT
```
Comment (Gandalf2k15) [inline code/modules/mob/mob_movement.dm]: //SKYRAT EDIT ADDITION BEGIN - GUNPOINT

File: code/modules/mob/mob_movement.dm [hunk 14]
```diff
@@ -449,6 +449,11 @@
...
+		if (HAS_TRAIT(src,TRAIT_NORUNNING))
+			to_chat(src, "You find yourself unable to run.")
+			return FALSE
+		//SKYRAT EDIT END
```
Comment (Gandalf2k15) [inline code/modules/mob/mob_movement.dm]: //SKYRAT EDIT ADDITION END

File: modular_skyrat/modules/gunpoint/readme.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+- /mob/living/ShiftMiddleClickOn() - overriden
+- /datum/radial_menu/proc/setup_menu() - modified
+- /obj/item/proc/attack_self() -modified
+- /obj/item/radio/talk_into() -modified
+- /obj/item/radio/headset/talk_into() - modified
+- /mob/living/carbon/human/examine() - modified
+- /mob/living/proc/update_mobility() - modified
+- /mob/living/proc/MobBump() - modified
+- /mob/proc/toggle_move_intent() - modified
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/gunpoint/readme.md]: Include file names

**Review by Gandalf2k15 — CHANGES_REQUESTED** [review]

**Comment by Azarak** [comment]
Adressed about everything

**Review by Gandalf2k15 — APPROVED** [review]
Final checks complete and passing. You are go for launch Major Tom.

**Review by Gandalf2k15 — APPROVED** [review]

**Review by Gandalf2k15 — APPROVED** [review]

**Pull request merged**